File: Luogu/Category/莫隊/P_4137_Rmq_Problem_mex.cpp
#include <bits/stdc++.h>
using namespace std;
#define endl '\n'

struct Query {
    int bid, l, r, qid;
    Query(int bid, int l, int r, int qid) : bid(bid), l(l), r(r), qid(qid) {}
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);

    int N, M;
    cin >> N >> M;
    vector<int> A(N);
    for (int i = 0; i < N; i++) cin >> A[i];

    // Mo's algorithm (只刪不加)
    // unordered_map<int, int> cnt0, cnt; // TLE
    vector<int> cnt0(N + 1), cnt(N + 1);
    int mex = 0, mex0 = 0;
    // 預處理整個陣列的 mex
    for (auto x : A)
        // 大於 N 的數字不會影響 mex
        if (x <= N) cnt0[x]++;
    while (cnt0[mex0] > 0) mex0++;

    auto del = [&](int x) {
        if (x <= N && --cnt[x] == 0) mex = min(mex, x);
    };

    // int BLK_SZ = sqrt(N);
    int BLK_SZ = ceil(N / sqrt(M * 2));

    vector<int> ans(M);
    vector<Query> queries;
    for (int qid = 0; qid < M; qid++) {
        int l, r;
        cin >> l >> r;
        l--;  // 0-indexed, [l, r]
        r--;
        // 大區間離線，確保 l 和 r 不在同一個 block 中
        if (r - l + 1 > BLK_SZ) queries.emplace_back(l / BLK_SZ, l, r, qid);
        // 小區間暴力
        else {
            fill(cnt.begin(), cnt.end(), 0);
            mex = 0;
            for (int i = l; i <= r; i++)
                if (A[i] <= N) cnt[A[i]]++;
            // mex 最多移動 BLK_SZ 次
            while (cnt[mex] > 0) mex++;
            ans[qid] = mex;
        }
    }

    // 離線排序
    sort(queries.begin(), queries.end(), [](const Query& a, const Query& b) {
        if (a.bid != b.bid) return a.bid < b.bid;
        return a.r > b.r;  // 右端點從大到小
    });

    int l = 0, r = 0;
    for (int i = 0; i < queries.size(); i++) {
        const auto& q = queries[i];
        int l0 = q.bid * BLK_SZ;
        // 遍歷到一個新的 block
        if (i == 0 || q.bid > queries[i - 1].bid) {
            // 左端點刪到 l0，這裡直接對 cnt0 進行操作
            for (; l < l0; l++)
                if (A[l] <= N && --cnt0[A[l]] == 0) mex0 = min(mex0, A[l]);
            // 複製 cnt0 到 cnt
            cnt = cnt0;
            mex = mex0;
            // 設定右端點
            r = N - 1;
        }
        // 右端點從 r 移動到 q.r (不包含 q.r)
        while (q.r < r) del(A[r--]);
        // 移動左端點前先備份 mex
        int tmp_mex = mex;
        // 左端點從 l0 移動到 q.l (不包含 q.l)
        for (int i = l0; i < q.l; i++) del(A[i]);
        ans[q.qid] = mex;

        // 回滾
        mex = tmp_mex;
        for (int i = l0; i < q.l; i++)
            if (A[i] <= N) cnt[A[i]]++;
    }
    for (const auto& x : ans) cout << x << endl;
    return 0;
}

File: Luogu/Tags/莫隊/P_4137_Rmq_Problem_mex.py
import math

import sys
input = lambda: sys.stdin.readline().strip()
print = lambda val: sys.stdout.write(str(val) + "\n")

N, Q = map(int, input().strip().split())
A = list(map(int, input().strip().split()))

cnt0 = [0] * (N + 1)
for x in A:
    if x <= N: cnt0[x] += 1
mex0 = 0
while cnt0[mex0] > 0:
    mex0 += 1

# 回滾莫隊（只減不加）
cnt = [0] * (N + 1)
mex = 0
def delete(x: int) -> None:
    global mex
    if x > N:  # 大於 N 的數字不會影響 mex
        return
    cnt[x] -= 1
    if cnt[x] == 0:
        mex = min(mex, x)

BLK_SZ = math.ceil(N / math.sqrt(Q * 2))
# BLK_SZ = math.sqrt(N)

ans = [-1] * Q
queries = []  # (bid, l, r, qid)
for qid in range(Q):
    l, r = map(int, input().split())
    l -= 1  # 0-indexed, [l, r]
    r -= 1
    # 大區間離線，確保 l 和 r 不在同一個 block 中
    if r - l + 1 > BLK_SZ:
        queries.append((l // BLK_SZ, l, r, qid))
        continue
    # 小區間暴力
    for i in range(N + 1):
        cnt[i] = 0
    mex = 0
    for i in range(l, r + 1):
        if A[i] <= N: cnt[A[i]] += 1
    while cnt[mex] > 0:  # 這裡最多移動 BLK_SZ 次
        mex += 1
    ans[qid] = mex

queries.sort(key=lambda x: (x[0], -x[2]))  # sort by block index and right index
for i, (bid, ql, qr, qid) in enumerate(queries):
    l0 = bid * BLK_SZ
    if i == 0 or bid > queries[i - 1][0]:  # 遍歷到一個新的 block
        # 複製 cnt0 到 cnt
        for i in range(N + 1):
            cnt[i] = cnt0[i]
        mex = mex0
        r = N - 1  # 右端點移動的起點
        # 左端點刪到 l0 (不包含 l0)
        for i in range(0, l0):
            delete(A[i])

    # 右端點從 r 移動到 qr (不包含 qr)
    while qr < r:
        delete(A[r])
        r -= 1
    tmp_mex = mex
    # 左端點從 l0 移動到 ql (不包含 ql)
    for i in range(l0, ql):
        delete(A[i])
    ans[qid] = mex

    # 回滾
    mex = tmp_mex
    for i in range(l0, ql):
        if A[i] <= N: cnt[A[i]] += 1

print('\n'.join(map(str, ans)))